source/scheduler.hpp

```diff
@@ -1,46 +1,71 @@
 #pragma once
 
-#include <memory>
 #include <queue>
 
 #include "event.hpp"
 
 namespace sim {
 
-struct EventComparator {
-    bool operator()(const std::unique_ptr<Event>& lhs,
-                    const std::unique_ptr<Event>& rhs) const {
-        return (*lhs.get()) > (*rhs.get());
-    }
-};
-
 // Scheduler is implemented as a Singleton class
 // which provides a global access to a single instance
-class Scheduler {
+template <typename TEvent>
+requires requires(const TEvent& e) {
+    { e.get_time() } -> std::same_as<Time>;
+}
+class SchedulerTemplate {
+    static_assert(
+        std::is_constructible_v<bool, decltype(std::declval<TEvent>() >
+                                               std::declval<TEvent>())>,
+        "Event must have operator <");
+    static_assert(std::is_invocable_v<TEvent>, "Event must be invocable");
+
 public:
     // Static method to get the instance
-    static Scheduler& get_instance() {
-        static Scheduler instance;
+    static SchedulerTemplate& get_instance() {
+        static SchedulerTemplate instance;
         return instance;
     }
 
-    void add(std::unique_ptr<Event> event);
-    void clear();  // Clear all events
-    bool tick();
-    Time get_current_time();
+    void add(const TEvent& event) { m_events.emplace(event); }
+
+    // Clear all events
+    void clear() {
+        m_events =
+            std::priority_queue<TEvent, std::vector<TEvent>, EventComparator>();
```

Comment:
Lets do it some faster
```suggestion
        std::priority_queue<TEvent, std::vector<TEvent>, EventComparator>().swap(m_events);
```

Reply:
Is it really faster?

Reply:
I little bit. Also it honestly frees the memory